bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs

```diff
@@ -0,0 +1,123 @@
+        Secret resource)
+    {
+        if (!_currentContext.AccessSecretsManager(resource.OrganizationId))
+        {
+            return;
+        }
+
+        switch (requirement)
+        {
+            case not null when requirement == SecretOperations.Create:
+                await CanCreateSecretAsync(context, requirement, resource);
+                break;
+            case not null when requirement == SecretOperations.Update:
+                await CanUpdateSecretAsync(context, requirement, resource);
+                break;
+            default:
+                throw new ArgumentException("Unsupported operation requirement type provided.", nameof(requirement));
+        }
+    }
+
+    private async Task CanCreateSecretAsync(AuthorizationHandlerContext context,
+        SecretOperationRequirement requirement, Secret resource)
+    {
+        var (accessClient, userId) = await _accessClientQuery.GetAccessClientAsync(context.User, resource.OrganizationId);
+        var project = resource.Projects?.FirstOrDefault();
+
+        if (project == null && accessClient != AccessClientType.NoAccessCheck)
+        {
+            return;
+        }
+
+        // All projects should be apart of the same organization
+        if (resource.Projects != null
+            && resource.Projects.Any()
+            && !await _projectRepository.ProjectsAreInOrganization(resource.Projects.Select(p => p.Id).ToList(),
+                resource.OrganizationId))
+        {
+            return;
+        }
+
+        var hasAccess = accessClient switch
+        {
+            AccessClientType.NoAccessCheck => true,
+            AccessClientType.User => (await _projectRepository.AccessToProjectAsync(project!.Id, userId, accessClient))
+                .Write,
+            AccessClientType.ServiceAccount => false,
+            _ => false,
+        };
+
+        if (hasAccess)
+        {
+            context.Succeed(requirement);
+        }
+    }
+
+    private async Task CanUpdateSecretAsync(AuthorizationHandlerContext context,
+        SecretOperationRequirement requirement, Secret resource)
+    {
+        var (accessClient, userId) = await _accessClientQuery.GetAccessClientAsync(context.User, resource.OrganizationId);
+
+        // All projects should be apart of the same organization
+        if (resource.Projects != null
+            && resource.Projects.Any()
+            && !await _projectRepository.ProjectsAreInOrganization(resource.Projects.Select(p => p.Id).ToList(),
+                resource.OrganizationId))
+        {
+            return;
+        }
+
+        bool hasAccess;
+
+        switch (accessClient)
+        {
+            case AccessClientType.NoAccessCheck:
+                hasAccess = true;
+                break;
+            case AccessClientType.User:
+                var newProject = resource.Projects?.FirstOrDefault();
+                var access = (await _secretRepository.AccessToSecretAsync(resource.Id, userId, accessClient)).Write;
```

Comment:
We removed a null check here, is that fine?

Reply:
In the original code:

```
          case AccessClientType.User:
                var oldProject = secret.Projects?.FirstOrDefault();
                var newProject = updatedSecret.Projects?.FirstOrDefault();
                var accessToOld = oldProject != null && (await _projectRepository.AccessToProjectAsync(oldProject.Id, userId, accessClient)).Write;
                var accessToNew = newProject != null && (await _projectRepository.AccessToProjectAsync(newProject.Id, userId, accessClient)).Write;
                return accessToOld && accessToNew;
```

We check access to the `oldProject` which is really the current project before processing the update and then a possible new project assignment.

I changed the logic around to instead:

  Does the user have access to the current state secret? 

```
var access = (await _secretRepository.AccessToSecretAsync(resource.Id, userId, accessClient)).Write;
```

And then check if they have access to a possible new project assignment.